File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	
	for n := 0; n < len(bytes); {
		m, err := socket.Write(bytes[n:])
		if err != nil {
			return err
		}
		n += m
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	n := 0

	for n < size {
		m, err := socket.Read(buff[n:])
		if err != nil {
			return nil, err
		}
		n += m
	}

	return buff, nil
	
}

File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(socket: socket.socket, size):

    bytes_recib = 0
    data = b""
    while bytes_recib < size:
        chunk = socket.recv(size - bytes_recib)
        if not chunk:
            raise Exception("Socket closed before receiving all data")
        data += chunk
        bytes_recib += len(chunk)
    return data


def send_all(socket: socket.socket, bytes):

    while len(bytes) > 0:
        bytes_sent = socket.send(bytes)
        bytes = bytes[bytes_sent:]

    return None